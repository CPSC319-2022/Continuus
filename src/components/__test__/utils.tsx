import React, { type PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { configureStore, type PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";

import type { AppStore, RootState } from "~/redux/store";
import { reducer } from "~/redux/slices";
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: reducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
