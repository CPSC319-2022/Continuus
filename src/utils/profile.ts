/**
 * Returns a path to the profile page for the given profile. If the argument is undefined, redirect to the home page.
 *
 * @param profile the user id of the account
 *
 * @return the profile path if the profile is a string, otherwise the home path
 */
export const userPathToProfile = (profile: string | undefined): string => {
    if (profile === undefined) {
        return '/';
    }

    return '/profile/' + encodeURIComponent(profile);
}
