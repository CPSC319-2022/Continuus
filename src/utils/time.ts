export const timeAgo = (inputDate: Date): string => {
  const now = new Date();
  const diffInSeconds = Math.max(Math.floor(
    (now.getTime() - inputDate.getTime()) / 1000
  ), 0);

  if (diffInSeconds < 60) {
    return `${diffInSeconds} ${diffInSeconds === 1 ? "second" : "seconds"} ago`;
  }

  const diffInMinutes = Math.floor(diffInSeconds / 60);
  if (diffInMinutes < 60) {
    return `${diffInMinutes} ${diffInMinutes === 1 ? "minute" : "minutes"} ago`;
  }

  const diffInHours = Math.floor(diffInMinutes / 60);
  if (diffInHours < 24) {
    return `${diffInHours} ${diffInHours === 1 ? "hour" : "hours"} ago`;
  }

  const diffInDays = Math.floor(diffInHours / 24);
  if (diffInDays < 30) {
    return `${diffInDays} ${diffInDays === 1 ? "day" : "days"} ago`;
  }

  const diffInMonths = Math.floor(diffInDays / 30);
  if (diffInMonths < 12) {
    return `${diffInMonths} ${diffInMonths === 1 ? 'month' : 'months'} ago`;
  }

  const diffInYears = Math.floor(diffInMonths / 12);
  return `${diffInYears} ${diffInYears === 1 ? 'year' : 'years'} ago`;
};
