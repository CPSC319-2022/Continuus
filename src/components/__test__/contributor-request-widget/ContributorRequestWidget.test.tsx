import { describe, it, expect, beforeEach, vi } from "vitest";
import { cleanup, render } from "@testing-library/react";
import { renderWithProviders } from "../utils";

vi.mock("next-auth/react", () => ({
  useSession: () => ({
    data: {
      expires: "1",
      user: { id: "test-user-id", email: "a", name: "hi", image: "c" },
    },
  }),
}));

const adminuser = {
  createdAt: new Date(),
  email: "email",
  id: "admin-id",
  image: "image",
  name: "admin",
  role: "ADMIN",
  updatedAt: new Date(),
};

const reader = {
  createdAt: new Date(),
  email: "email",
  id: "reader-id",
  image: "image",
  name: "reader",
  role: "READER",
  updatedAt: new Date(),
};

describe("Contributor Request Menu: Snapshot", () => {
  beforeEach(() => {
    cleanup();
    vi.clearAllMocks();
    vi.resetModules();
  });

  const setup = (user = adminuser) => {
    vi.doMock("~/utils/api", () => ({
      api: {
        user: {
          currentUser: {
            useQuery: () => ({
              data: user,
              isLoading: false,
            }),
          },
          updateOne: {
            useMutation: vi.fn(),
          },
        },
        contributorRequest: {
          getAll: {
            useQuery: () => ({
              data: [
                {
                  id: "contribreq-id-1",
                  createdAt: new Date(),
                  user: {
                    id: "user-id-1",
                    name: "test user 1",
                    image: "image",
                  },
                },
                {
                  id: "contribreq-id-2",
                  createdAt: new Date(),
                  user: {
                    id: "user-id-2",
                    name: "test user 2",
                    image: "image",
                  },
                },
              ],
              isLoading: false,
            }),
          },
          delete: {
            useMutation: () => ({
              mutate: vi.fn(),
              mutateAsync: vi.fn(),
            }),
          },
        },
        useContext: () => ({
          contributorRequest: {
            invalidate: vi.fn(),
          },
        }),
      },
    }));
  };

  it("render as admin", async () => {
    setup(adminuser);
    const { ContributorRequestWidget } = await import(
      "~/components/contributor-request-widget/ContributorRequestWidget"
    );
    const body = renderWithProviders(<ContributorRequestWidget />).baseElement;
    expect(body).toMatchSnapshot();
  });

  it("render as reader", async () => {
    setup(reader);
    const { ContributorRequestWidget } = await import(
      "~/components/contributor-request-widget/ContributorRequestWidget"
    );
    const body = renderWithProviders(<ContributorRequestWidget />).baseElement;
    expect(body).toMatchSnapshot();
  });
});
