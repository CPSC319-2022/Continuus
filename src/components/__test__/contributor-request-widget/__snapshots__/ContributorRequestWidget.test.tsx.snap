// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Contributor Request Menu: Snapshot > render as admin 1`] = `
<body>
  <div>
    <div>
      <div
        class="hover:cursor-pointer hover:fill-highlight-green transition-all"
      >
        <svg
          height="24"
          viewBox="0 96 960 960"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M160 856v-80h80V496q0-83 50-147.5T420 264v-28q0-25 17.5-42.5T480 176q25 0 42.5 17.5T540 236v28q80 20 130 84.5T720 496v280h80v80H160Zm320-300Zm0 420q-33 0-56.5-23.5T400 896h160q0 33-23.5 56.5T480 976ZM320 776h320V496q0-66-47-113t-113-47q-66 0-113 47t-47 113v280Z"
          />
        </svg>
      </div>
    </div>
  </div>
</body>
`;

exports[`Contributor Request Menu: Snapshot > render as reader 1`] = `
<body>
  <div />
</body>
`;
